### ISSUE TITLE
Autograd always backpropagates to arithmetic operations when there is an unused input

### ISSUE THREAD

**Issue by patrick-kidger** [pytorch/pytorch#39784]
## 🐛 Bug

If:
- a simple arithmetic operation (add, multiply) is used in the computation graph
- AND `autograd.grad` is called wrt an unused input

then:
- autograd will include this arithmetic operation (and everything leading back to it) in its backward pass.

Remarkably, this occurs even if `allow_unused=False`. The backpropagation is performed first, and the error about unused inputs is thrown afterwards!

I've not tested every operation (subtraction, division, etc.), but I can confirm that this bug doesn't occur with certain other operations (clone, relu, view).

## To Reproduce

```python
import torch

class Print(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x, op, final, unused):
        ctx.op = op
        ctx.final = final
        ctx.unused = unused
        return x

    @staticmethod
    def backward(ctx, x):
        print('op={}, final={}, unused={} has a bug'.format(ctx.op, ctx.final, ctx.unused))
        return x, None, None, None

def f(op, final, unused):
    x = torch.rand(1, requires_grad=True)
    if op == 'mul':
        x = 2 * x
    elif op == 'add':
        x = 2 + x
    elif op == 'relu':
        x = x.relu()
    elif op == 'clone':
        x = x.clone()
    elif op == 'view':
        x = x.view(-1)
    x = Print.apply(x, op, final, unused)

    y = torch.rand(1, requires_grad=True)

    if final:
        x = y + x

    try:
        torch.autograd.grad((x,), (y,), allow_unused=unused)
    except RuntimeError:
        pass


f('mul', False, True)
f('add', False, True)
f('relu', False, True)
f('clone', False, True)
f('view', False, True)

f('mul', True, True)
f('add', True, True)
f('relu', True, True)
f('clone', True, True)
f('view', True, True)

f('mul', True, False)
f('add', True, False)
f('relu', True, False)
f('clone', True, False)
f('view', True, False)

f('mul', False, False)
f('add', False, False)
f('relu', False, False)
f('clone', False, False)
f('view', False, False)
```

Prints:
```
op=mul, final=False, unused=True has a bug
op=add, final=False, unused=True has a bug
op=mul, final=False, unused=False has a bug
op=add, final=False, unused=False has a bug
```

## Expected behavior

Nothing to be printed at all!

## Environment

 - PyTorch Version (e.g., 1.0): 1.5
 - OS (e.g., Linux): Ubuntu 18.04 LTS
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.8.2

## Additional context

This came up in a context where I'm handling the backward computation quite carefully - and this bug is currently producing several-order-of-magnitude slowdowns because the autograd engine is backprop'ing through very expensive parts of the graph that it really shouldn't be touching.

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by albanD** [comment]
Hi,

This is indeed our python engine that ignores inputs that require grad but were never used [here](https://github.com/pytorch/pytorch/blob/019eeb3183e5787553a32bfdc61fbe32e9513b0f/torch/csrc/autograd/python_engine.cpp#L195). I guess this should through an error.
Making sure that `y` in your code sample is actually used (add `_ = y.clone()` after creating it) fixes the issue. Is that a good enough temporary workaround for you?

We should properly fix the engine to handle such case though.

**Comment by SsnL** [comment]
I don't think it should throw an error when `allow_unused=True` though

**Comment by albanD** [comment]
Yes, of course. I haven't looked into much detail. But the root cause is that this empty edge is not properly handled by the discovery code. We should just fix that.

**Comment by patrick-kidger** [comment]
Curious that it triggers for `+` and `*` but not `relu` etc. though. Thanks for the suggestion on a workaround. So far I've been doing `x = x + 0 * y`, which also works. :)

**Comment by albanD** [comment]
You don't even need `y` to be in the same graph. Just doing any operation with in will fix the issue (because it will materialize the gradient accumulator for it).
